### ISSUE TITLE
Integrated device not shown up "fast enough" on startup

### ISSUE THREAD

**Issue by alosarjos** [aome510/spotify-player#79]
Lately I've noticed that spotify-player is starting without picking the default device (Which is itself). In the config file is like this:

```
default_device = "artorias"

[device]
name = "artorias"
device_type = "computer"
volume = 100
bitrate = 320
audio_cache = false
```

So whenever I open the player, I have to choose a device manually (There it's listed) and then I can play music. Is there some kind of race condition or something for picking the default device maybe?

**Comment by aome510** [comment]
> Lately I've noticed that spotify-player is starting without picking the default device (Which is itself)

@alosarjos can you share one example log of such case?

> So whenever I open the player, I have to choose a device manually (There it's listed) and then I can play music. 

Does the app display something like *"No playback found. Please make sure there is a running Spotify client and try to connect to it using the `SwitchDevice` command."*?

**Comment by alosarjos** [comment]
Sure, I remove some keys/ids I saw since I wasn't sure those should be "public" just in case

And yes, that is exactly the message it says.

[spotify-player-22-09-11-19:59.log](https://github.com/aome510/spotify-player/files/9543220/spotify-player-22-09-11-19.59.log)

**Comment by alosarjos** [comment]
I'm assuming the issue comes from this info lines:

```
2022-09-11T17:59:51.802311Z  INFO spotify_player: No playing device found on startup, trying to connect to an available device
2022-09-11T17:59:52.105250Z  INFO spotify_player::client: Available devices: []
```

Where, for some reason the device is not showing up fast enough?

**Comment by alosarjos** [comment]
The device is correctly initialized, it shows up if I open de device selector and it's also available on the phone app.

**Comment by aome510** [comment]
Thanks for the log and further investigation. Look like a known issue to me.

> Where, for some reason the device is not showing up fast enough?

Yes, the reason is the new device is not shown up fast enough when calling `get_device` API for available devices, at least based on my understanding. I've encountered this issue a few times before. I also made some changes to make it happen less frequently, but I guess they're not enough. 

Related discussion: https://github.com/aome510/spotify-player/issues/69#issuecomment-1211556708. 

I'll change the title to better reflect the issue. In the meantime, I'll think more about ways to improve this startup process.

**Comment by elkrien** [comment]
Have the same problem with new version. I had this problem in the past where there was no option for default device. Then after implementing this option it was ok, now after update player is not reading my config. When I open player it starts playing on first device on the list (in my situation is Sony TV). It's very frustrating because it turns on tv every time I start player, then I need to change device manually and go downstairs to turn off tv ;)

**Comment by aome510** [comment]
> It's very frustrating because it turns on tv every time I start player, then I need to change device manually and go downstairs to turn off tv ;)

That's unfortunate it happens again. I'm currently implementing a proper fix (hope so) for this issue. Will ping you guys once it's ready

**Comment by alosarjos** [comment]
@aome510 Sure thing. If you ping me with a branch or a commit I can give it a try before tagging the release if you need it. ;)

**Comment by aome510** [comment]
@alosarjos @elkrien implemented the fix in https://github.com/aome510/spotify-player/pull/81. Can you check it out and try if the new changes fix this issue?

Basically, what I did is to add retries when connecting to the integrated device. That means it ensures that the integrated device is picked on startup but with delay. 

Regarding the issue after update, I don't particularly know why the integrated device registration is much slower (maybe because of third-party library changes). Let's leave that to a separated issue and do further investigation if the new fix is not enough.

**Comment by alosarjos** [comment]
@aome510 For my is having a strange behaviour, I keep opening and closing the app and I've seen 3 things happening in all those tries:

1. Takes a bit of time, but finds the device
2. Keeps failing when looking for the device
3. Sometimes the playback info is shown (So, device is found) for a split second and then goes back to No playback found.

**Comment by elkrien** [comment]
@aome510 It starts a lot slower now but works fine - no more turning on my TV :)

**Comment by aome510** [comment]
> 2. Keeps failing when looking for the device
> 3. Sometimes the playback info is shown (So, device is found) for a split second and then goes back to No playback found.

Interesting. Will take a look 👀

**Comment by aome510** [comment]
> Sometimes the playback info is shown (So, device is found) for a split second and then goes back to No playback found.

I also experience the same thing occasionally. Looking into the log in that case, it seems that the app displays the playback from the previous run, whose device is shut down but not updated in the Spotify server. Because there is already a running playback at startup, it doesn't try to connect to one. Then, when the shutdown event is known, the playback becomes empty.

> Keeps failing when looking for the device

I think this is also because of the same thing as above. The app knows there is one running on startup, so it doesn't look for one.

**Comment by aome510** [comment]
Should be fixed by #81. I'll close this for now. Feel free to re-open it if encounter any related issues.
